### ISSUE TITLE
ImportOrder: Check that import groups aren't separated internally

### ISSUE THREAD

**Issue by robertwhitebit** [checkstyle/checkstyle#2143]
Currently, the ImportOrder check ensures that imports are grouped in a specific order. These groups shouldn't be separated internally as well.

e.g.

``` java
import com.puppycrawl.tools.checkstyle.api.Check;

import com.puppycrawl.tools.checkstyle.api.DetailAST;
import com.puppycrawl.tools.checkstyle.api.TokenTypes;
```

The empty line between Check and DetailAST should be removed. Automatically importing classes will never add such a newline. The Android Studio formatter also removes the empty line.
##

**Comment by sshashank124** [comment]
Until what point should they be grouped. Would you say that these two statements should also be grouped or only the innermost ones:

```
import org.apache.commons.beanutils.ConversionException;
import org.apache.commons.lang3.StringUtils;
```

**Comment by mkordas** [comment]
@sshashank124, I believe that we should just not allow not explicitly allowed newlines in imports to match major IDEs formatting/optimise imports rules

**Comment by robertwhitebit** [comment]
@sshashank124: This is a good question!
@mkordas: Exactly.

I tested the Eclipse and Android Studio (IntelliJ) formatter settings. You can configure how and at what level you define groups.
If you're using "java, com" then the formatter of the tested IDEs will group at the first level of the import.

```java
import java.io.IOException;

import com.oracle.net.Sdp;
import com.oracle.nio.BufferSecretsPermission;
```
Inserting a newline between the two com.* imports and running the formatter/optimize imports will always remove the newline automatically.

However, if you are using something like "java, com.oracle.nio, com.oracle.net" then the result of the formatter is:

```java
import java.io.IOException;

import com.oracle.nio.BufferSecretsPermission;

import com.oracle.net.Sdp;
```

**Comment by sshashank124** [comment]
So then does it make sense to leave the level of grouping up to the individual or were you thinking of a predefined set of grouping rules

**Comment by robertwhitebit** [comment]
Hopefully I didn't misinterpret the question. :worried:

The level of the grouping is already up to the user. You can set the groups property to e.g. `"java,com.oracle.nio,com.oracle.net"`. So the second code example in the previous post is handled by checkstyle already.

Imo, the only thing that has to be done on the user side, is, to add a flag (true/false) if it's allowed to separate these groups internally.

So, I was just thinking of a flag. :wink:

**Comment by sshashank124** [comment]
Ah I see. My bad. Yeah that's a good idea. So now, will someone delegate this to someone to implement if it is finally decided to be added?

**Comment by mkordas** [comment]
@sshashank124, anyone can take it and submit a PR. All contributions are welcome, also from you guys :)

**Comment by sshashank124** [comment]
@mkordas, nice. I don't know enough about checkstyle at the moment to contribute. Definitely on future issues

**Comment by robertwhitebit** [comment]
An interesting case is a comment in a group. Like

```java
import com.puppycrawl.tools.checkstyle.api.Check;
// import com.puppycrawl.tools.checkstyle.api.DetailAST;
import com.puppycrawl.tools.checkstyle.api.TokenTypes;
```

Eclipse and IntelliJ move or remove them. See https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/373 for more details. I think that in these cases, the formatter does more than it should. But I justed wanted to share this information here as well.

**Comment by romani** [comment]
@linelect , can you help us to resolve this issue?

**Comment by romani** [comment]
fix is merged

**Comment by jonmbake** [comment]
@romani Shouldn't this only be enforced when `ImportOrder. separated=true`?

I am attempting to update from CheckStyle 6.14.1 to 8.7. I have the following CheckStyle rule specified:
```
<module name="ImportOrder">
    <property name="separated" value="false"/>
</module>
```

When running 6.14.1 CheckStyle, there are no violations. When running 8.7, there are hundreds of "Extra separation in import group before" violations. For example,
```
import javax.naming.NamingException;

import org.slf4j.Logger;

public class Foo {
}
```
Passed in 6.14.1, but now fails. This seems like a regression to me.

**Comment by romani** [comment]
@jonmbake , please open new issue, smth is brocken.
if you represent opensource code please provide link to project and how to run checkstyle - we will recheck that after fix you have no problems and we will add you to CI process to avoid regression.

**Comment by robertwhitebit** [comment]
@linelect, Awesome! Thank you very much for fixing it! :smiley:
And, as always, also a big thanks to the Checkstyle team!
